automation_tools/scripts/ga-migration/main.py

```diff
@@ -0,0 +1,284 @@
+    # If the file exists
+    if os.path.isfile(filepath):
+        # And the line is not already there
+        if not file_contains(term, filepath):
+            append_to_file(term, filepath)
+        else:
+            logging.info("SKIPPED TASK. Line already there. ")
+    else:
+        logging.info("SKIPPED TASK. No %s found" % filepath)
+
+
+def replace_simple(text, replacing, filepath):
+    """
+    Replaces every match of a string with another in the specified file
+    """
+    logging.info(
+        "TASK: Simple replacing %s with %s in %s" % (text, replacing, filepath)
+    )
+    if os.path.isfile(filepath):
+        logging.info("Found %s" % filepath)
+        # TODO: expose number of matches
+        with fileinput.FileInput(filepath, inplace=True, backup=".bak") as file:
+            for line in file:
+                print(line.replace(text, replacing), end="")
+    else:
+        logging.info("No %s found" % filepath)
+
+
+def replace_regex(regex, output, filepath):
+    """
+    Replaces every match of a string with another in the specified file
+    """
+    logging.info(
+        "TASK: RegEx replacing %s with %s in %s" % (regex, output, filepath)
+    )
+    if os.path.isfile(filepath):
+        logging.info("Found %s" % filepath)
+        # TODO: expose number of matches
+        with fileinput.FileInput(filepath, inplace=True, backup=".bak") as file:
+            for line in file:
+                print(re.sub(regex, output, line), end="")
+    else:
+        logging.info("SKIPPED TASK. No %s found" % filepath)
+
+
+def download_file(url, destination):
+    # Get path
+    dirname = os.path.dirname(os.path.realpath(destination))
+    if not os.path.exists(dirname):
+        os.makedirs(dirname)
+    print(dirname)
+    r = requests.get(url, allow_redirects=True)
+    open(destination, "wb").write(r.content)
+
+
+def replace_list(filepath, regex, to_remove, to_add, var_name):
+    """
+    Given a python file, look for the "var_name" using "regex" and:
+    - remove any occurence of the elements from the "to_remove" list
+        partial matches allowed, e.g. pytest-cov will remove pytest-cov>=0.0.1
+    - add the elements from the "to_add" list
+    Write the changes to "var_name" variable in the original file
+    """
+
+    with open(filepath, "r") as f:
+        contents = f.read()
+
+    # Search the list in the file contents
+    m = re.search(regex, contents)
+
+    # Group zero matches the whole assignment,
+    # We need the right part of the assignment (Group 1)
+    matched_list_str = m.group(1)
+
+    # Deserialize it
+    parsedlist = ast.literal_eval(matched_list_str)
+
+    # For each element to remove, check if it's present in the list
+    for element_to_remove in to_remove:
+        for element in parsedlist:
+            # If it's part of an element (partial matches allowed), remove it
+            if element_to_remove in element:
+                parsedlist.remove(element)
```

Comment:
Major: I think these nested loops are both inefficient and are causing troubles in some corner cases.
For instance, after running the script in `invenio-i18n`, the results is:

```python
tests_require = [
    "invenio-accounts[sqlite]>=1.1.3",
    "invenio-assets>=1.0.0",
    "isort>=4.3.0",
    "pydocstyle>=1.0.0",
    "pytest>=3.8.0,<5.0.0",
    "pytest-invenio>=1.4.0"
]
```

Note that `pytest` remains there even though it is part of `to_remove`. I see two big problems:
1. Relying on `if element_to_remove in element` is a very week condition as e.g. evaluates to `True` for all `pytest-*` packages when `element_to_remove` is `pytest`.
2. It's not considered a good practice removing items from lists you're iterating over. This is causing strange behaviors like the one described above where it's basically skipping one item because other gets deleted and probably the internal indexes are getting crazy (not fully sure but this is what I could see while debugging, you can try it yourselves).

I wonder if this can be better done with sets. In that case, we should extract the versions into a dict, perform the `difference` between sets, and then attach the versions again to the result.

Reply:
Removed the nested loops. Now the match is exact and the removal/add is not done in-place. https://github.com/inveniosoftware/automation-tools/pull/2/commits/7b947a0eab41c7f8e7059c4ffbf8aaa88f77bd1c